napari/_qt/dialogs/activity_dialog.py

```diff
@@ -0,0 +1,217 @@
+
+        self._activityLayout = QVBoxLayout()
+        self._activityLayout.addStretch()
+        self._baseWidget.setLayout(self._activityLayout)
+        self._baseWidget.layout().setContentsMargins(0, 0, 0, 0)
+
+        self._scrollArea = QScrollArea()
+        self._scrollArea.setWidgetResizable(True)
+        self._scrollArea.setWidget(self._baseWidget)
+
+        self._titleBar = QLabel()
+
+        title = QLabel('activity', self)
+        title.setObjectName('QtCustomTitleLabel')
+        title.setSizePolicy(
+            QSizePolicy(QSizePolicy.Policy.Maximum, QSizePolicy.Policy.Maximum)
+        )
+        line = QFrame(self)
+        line.setObjectName("QtCustomTitleBarLine")
+        titleLayout = QHBoxLayout()
+        titleLayout.setSpacing(4)
+        titleLayout.setContentsMargins(8, 1, 8, 0)
+        line.setFixedHeight(1)
+        titleLayout.addWidget(line)
+        titleLayout.addWidget(title)
+        self._titleBar.setLayout(titleLayout)
+
+        self._baseLayout = QVBoxLayout()
+        self._baseLayout.addWidget(self._titleBar)
+        self._baseLayout.addWidget(self._scrollArea)
+        self.setLayout(self._baseLayout)
+
+    def add_progress_bar(self, pbar, nest_under=None):
+        """Add progress bar to the activity_dialog, making ProgressBarGroup if needed.
+
+        Check whether pbar is nested and create ProgressBarGroup if it is, removing
+        existing separators and creating new ones. Show and start inProgressIndicator
+        to highlight to user the existence of a progress bar in the dock even when
+        the dock is hidden.
+
+        Parameters
+        ----------
+        pbar : ProgressBar
+            progress bar to add to activity dialog
+        nest_under : Optional[ProgressBar]
+            parent progress bar pbar should be nested under, by default None
+        """
+        if nest_under is None:
+            self._activityLayout.addWidget(pbar)
+        else:
+            # this is going to be nested, remove separators
+            # as the group will have its own
+            parent_pbar = nest_under._pbar
+            current_pbars = [parent_pbar, pbar]
+            remove_separators(current_pbars)
+
+            parent_widg = parent_pbar.parent()
+            # if we are already in a group, add pbar to existing group
+            if isinstance(parent_widg, ProgressBarGroup):
+                nested_layout = parent_widg.layout()
+            # create ProgressBarGroup for this pbar
+            else:
+                new_group = ProgressBarGroup(nest_under._pbar)
+                new_group.destroyed.connect(self.maybe_hide_progress_indicator)
+                nested_layout = new_group.layout()
+                self._activityLayout.addWidget(new_group)
+            new_pbar_index = nested_layout.count() - 1
+            nested_layout.insertWidget(new_pbar_index, pbar)
+
+        # show progress indicator and start gif
+        self._toggleButton._inProgressIndicator.movie().start()
```

Comment:
this bit -- can it be in a new thread? @tlambert03 might have advice on how to do that cleanly...

Reply:
I think we could definitely put the indicator in its own thread, but I think it's worth noting here that the reason it's so jarringly slow in the segmentation example, and in the minimal examples, is because I've added `sleep` calls to make the processing a bit slower so the user can see the progress bars. I struggled to come up with an example that took long enough to process to really highlight the progress bar, but was still quick enough to be an example. I think the only one we've really achieved that with is the random number generating in the indeterminate example.